flax/errors.py

```diff
@@ -297,5 +323,263 @@ def __call__(self, x):
+
+    class Foo(nn.Module):
+      features: int = 6
+
+      def setup(self):
+        self.features = 3  # <-- ERROR
+
+      def __call__(self, x):
+        return nn.Dense(self.features)(x)
+
+    variables = SomeModule().init(random.PRNGKey(0), jnp.ones((1, )))
+
+  Instead, these attributes should be set when initializing the Module::
+
+    class Foo(nn.Module):
+      features: int = 6
+
+      @nn.compact
+      def __call__(self, x):
+        return nn.Dense(self.features)(x)
+
+    variables = SomeModule(features=3).init(random.PRNGKey(0), jnp.ones((1, )))
+        
+  """
+  def __init__(self):
+    super().__init__(f'Module construction attributes are frozen.')
+
+
+class SetAttributeFrozenModuleError(FlaxError):
+  """
+  You can only assign Module attributes to ``self`` inside
+  :meth:`Module.setup() <flax.linen.Module.setup>`. Outside of that method, the
+  Module instance is frozen (i.e., immutable). This behavior is similar to
+  frozen Python dataclasses.
+  
+  For instance, this error is raised in the following case::
+
+    class SomeModule(nn.Module):
+      @nn.compact
+      def __call__(self, x, num_features=10):
+        self.num_features = num_features  # <-- ERROR!
+        x = nn.Dense(self.num_features)(x)
+        return x
+
+    s = SomeModule().init(random.PRNGKey(0), jnp.ones((5, 5)))
+
+  Similarly, the error is raised when trying to modify a submodule's attributes
+  after constructing it, even if this is done in the ``setup()`` method of the
+  parent module::
+
+    class Foo(nn.Module):
+        def setup(self):
+          self.dense = nn.Dense(features=10)
+          self.dense.features = 20  # <--- This is not allowed
+        
+        def __call__(self, x):
+          return self.dense(x)
+  """
+  def __init__(self, module_cls, attr_name, attr_val):
+    super().__init__(f'Can\'t set {attr_name}={attr_val} for Module of type '
+                    f'{module_cls}: Module instance is frozen outside of '
+                     'setup method.')
+
+
+class MultipleMethodsCompactError(FlaxError):
+  """
+  The ``@compact`` decorator may only be added to at most one method in a Flax
+  module. In order to resolve this, you can:
+  
+  * remove ``@compact`` and define submodules and variables using 
+    :meth:`Module.setup() <flax.linen.Module.setup>`.
+  * Use two separate modules that both have a unique ``@compact`` method.
+  """
+  def __init__(self):
+    super().__init__(f'Only one method per class can be @compact')
+
+class ReservedModuleAttributeError(FlaxError):
+  """
+  This error is thrown when creating a Module that is using reserved attributes.
+  The following attributes are reserved:
+  
+  * ``parent``: The parent Module of this Module.
+  * ``name``: The name of this Module.
+  """
+  def __init__(self, annotations):
+    super().__init__(f'properties `parent` and `name` are reserved: '
+                     f'{annotations}')
+
+
+class InitModuleInvalidRngsError(FlaxError):
```

Comment:
Same here, the user knows they are in init by looking at the stack trace. So I think instead of this class you can just raise `InvalidRngError`?

Reply:
(Is there also `ApplyRngError`?)